firedrake/extrusion_numbering.pyx

```diff
@@ -0,0 +1,762 @@
+    """
+    cdef:
+        PETSc.DM dm
+        DMLabel label
+        PetscInt p, point, layers, i, j, dimension
+        numpy.ndarray[PetscInt, ndim=2, mode="c"] nodes
+        numpy.ndarray[PetscInt, ndim=2, mode="c"] layer_extents = mesh.layer_extents
+        numpy.ndarray[PetscInt, ndim=2, mode="c"] stratum_bounds
+        numpy.ndarray[PetscInt, ndim=1, mode="c"] node_classes
+        numpy.ndarray[PetscInt, ndim=1, mode="c"] indices
+
+    nodes = numpy.asarray(nodes_per_entity, dtype=IntType)
+
+    node_classes = numpy.zeros(4, dtype=IntType)
+
+    dm = mesh._plex
+    dimension = dm.getDimension()
+    stratum_bounds = numpy.zeros((dimension + 1, 2), dtype=IntType)
+    for i in range(dimension + 1):
+        stratum_bounds[i, :] = dm.getDepthStratum(i)
+
+    for i, lbl in enumerate(["op2_core", "op2_non_core",
+                             "op2_exec_halo", "op2_non_exec_halo"]):
+        if dm.getStratumSize(lbl, 1) < 1:
+            continue
+        indices = dm.getStratumIS(lbl, 1).indices
+        for p in range(indices.shape[0]):
+            point = indices[p]
+            layers = layer_extents[point, 1] - layer_extents[point, 0]
+            for j in range(dimension + 1):
+                if stratum_bounds[j, 0] <= point < stratum_bounds[j, 1]:
+                    node_classes[i] += nodes[j, 0]*layers + nodes[j, 1]*(layers - 1)
+                    break
+
+    return numpy.cumsum(node_classes)
+
+
+@cython.wraparound(False)
+def entity_layers(mesh, height, label=None):
+    """Compute the layers for a given entity type.
+
+    :arg mesh: the extruded mesh to compute layers for.
+    :arg height: the height of the entity to consider (in the DMPlex
+       sense). e.g. 0 -> cells, 1 -> facets, etc...
+    :arg label: optional label to select some subset of the points of
+       the given height (may be None meaning select all points).
+
+    :returns: a numpy array of shape (num_entities, 2) providing the
+       layer extents for iteration on the requested entities.
+    """
+    cdef:
+        PETSc.DM dm
+        DMLabel clabel = NULL
+        numpy.ndarray[PetscInt, ndim=2, mode="c"] layer_extents
+        numpy.ndarray[PetscInt, ndim=2, mode="c"] layers
+        PetscInt f, p, i, hStart, hEnd, pStart, pEnd
+        PetscInt point, offset
+        const PetscInt *renumbering
+        PetscBool flg
+
+    dm = mesh._plex
+
+    hStart, hEnd = dm.getHeightStratum(height)
+    if label is None:
+        size = hEnd - hStart
+    else:
+        size = dm.getStratumSize(label, 1)
+
+    layers = numpy.zeros((size, 2), dtype=IntType)
+
+    layer_extents = mesh.layer_extents
+    offset = 0
+    if label is not None:
+        CHKERR(DMGetLabel(dm.dm, <char *>label, &clabel))
+        CHKERR(DMLabelCreateIndex(clabel, hStart, hEnd))
+    pStart, pEnd = dm.getChart()
+    CHKERR(ISGetIndices((<PETSc.IS?>mesh._plex_renumbering).iset, &renumbering))
+    for p in range(pStart, pEnd):
+        point = renumbering[p]
```

Comment:
This is the translation from DMPlex point numbering to Firedrake numbering.  By writing this generically we don't have to build a Section that gives us the renumbering for each entity type.